### ISSUE TITLE
The result of the failed optimization is either Input bitseq incorrect or RECHECK

### ISSUE THREAD

**Issue by lrlrl** [open-s4c/vsyncer#59]
When I want to explore whether the optimization suggestions given by vsyncer are based on the implementation of the synchronization primitive, or whether the optimization suggestions of the primitive are based on the call in the invocation scenario, I try to define multiple locks such as lock1 and lock2, and pass the locks as variables to the lock() and unlock() functions, and call the run() function, but the result is  stuck in '**RECHECK**'.
![RECHECK](https://github.com/user-attachments/assets/6b2baefb-41c2-42c4-ad75-4ed88f64fb9a)
In addition, when I implement multiple types of locks in client code and call their respective lock() and unlock() in run(), the result is **Input bitseq incorrect**.
![Input bitseq incorrect](https://github.com/user-attachments/assets/07640a48-f9fa-4f0f-bf77-77e007732258)
The code is placed in a compressed package for reference.
[vsyncer_problem.zip](https://github.com/user-attachments/files/17714301/vsyncer_problem.zip)

**Comment by lrlrl** [comment]
I'm only using the Vsyncer default configuration during the optimization process，and optimize with 

`vsyncer optimize -A -1 examples/xxxlock.c`

depending on the configuration, the Dat3M backend will also check the case where pthread_create fails. But Vsyncer's default model checker is GenMC, I don’t know whether this configuration is used by VSyncer.

**Comment by db7** [comment]
Another thing: we never really ran genmc on RISC-V. You don't need to run vsyncer inside your target machine. You can run that on a normal Intel CPU if you prefer. The model checking happens at the builtin atomic level, which is mapped by the compiler to the target architecture when the real program is compiled.

**Comment by lrlrl** [comment]
Yes, I modified the lock for examples, but I also had no problem running examples/ttasklock.c locally. In screenshot one, my bitseq is shorter than yours because I ran lock() and unlock() in it multiple times. If you run lock() and unlock() only once, you can get optimization results. Screenshot 2 may be due to the fact that multiple locks are implemented in the same .c file.

When I run it in Ubuntu22, I still get the above error.

**Comment by lrlrl** [comment]
The code is placed in a compressed package for reference.
[vsyncer_problem.zip](https://github.com/user-attachments/files/17715403/vsyncer_problem.zip)

**Comment by db7** [comment]
Sorry for the long delay. It seems your assertion is wrong at the end of the program.

```
    assert(x == y);
    assert(x == NTHREADS * 2);
```

That is because, AFAICT, the threads increment `x` and `y` only once.

```
void *
run(void *_)
{

    ttas_lock();
    cas_lock();
    x++;
    y++;
    cas_unlock();
    ttas_unlock();


    // cas_lock();
    // x++;
    // y++;
    // cas_unlock();

    return 0;
}
```

Also note that in this code, if you optimize it, `vsyncer` will tell you that the `cas_lock` doesn't need any barriers (because it is protected by the `ttas_lock`.

Please let me know if that clarify your issue, and sorry again for the long delay.

**Comment by lrlrl** [comment]
Yes, you're right. When I modified the assertion `assert(x == NTHREADS);`, the problem was resolved.But when I optimize it，`vsyncer` tell me that the **`ttas_lock`** doesn't need any barriers.

![Image](https://github.com/user-attachments/assets/6f00f574-fa05-40f4-9641-f8aed29a1172)

`l2` is the variable of `cas_lock`. However, this has nothing to do with the issue, the modified code can be successfully optimized, and the issue can be closed.

Thank you for your help.

**Comment by db7** [comment]
Yes, that is what I meant before. In your test case you are protecting the critical section with 2 locks. One of them is unnecessary, hence the optimizer will get rid of the barriers. In my comment above, I thought it would eliminate the barriers of the caslock, but it seems it chose the barriers of ttaslock.

The optimization is for the specific client code (ie, test case) you have. The problem is that your client code is too simple for the case of two locks. Either you create individual test cases for them, or you need to make both locks to be essential in the current test case.
